### PR TITLE
outline: fix broken links to public assets

### PR DESCRIPTION
**Pull request by weiren2**
Static images in Outline are no longer accessible after commit 06a70ad. Links to all favicons and static images are broken and redirected to an HTML page.

The public assets has been moved to `$out/share/outline/build` (in order to fix [email notification issue](https://github.com/NixOS/nixpkgs/commit/06a70adfa9c6f303306a595bb675465e40080311)), but those are no longer at the [expected path](https://github.com/outline/outline/blob/v0.67.2/server/routes/index.ts#L32) in Outline. This patch lets Outline find the correct path to its public assets again.
With this patch, the links to these static images are working (at least in my local tests). 

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by NickCao** [comment]
So why either the email template or the public assets breaks? Are we incorrectly setting the working directory?

**Comment by NickCao** [comment]
Can we change: https://github.com/NixOS/nixpkgs/blob/042598f1989709751cab749a94dffdba9e926593/nixos/modules/services/web-apps/outline.nix#L784

To `${cfg.package}/share/outline`

Move everything in `build` up a level and revert https://github.com/NixOS/nixpkgs/commit/06a70adfa9c6f303306a595bb675465e40080311.

**Comment by weiren2** [comment]
> So why either the email template or the public assets breaks? Are we incorrectly setting the working directory?

I'm not entirely sure. I think it's likely due to the working directory different from upstream.

Since I don't use any SMTP service in Outline, it's hard for me to test the issue related to email (as mentioned 06a70ad). Maybe @xanderio can elaborate more on this.

> Can we change:
> 
> https://github.com/NixOS/nixpkgs/blob/042598f1989709751cab749a94dffdba9e926593/nixos/modules/services/web-apps/outline.nix#L784
> 
> To `${cfg.package}/share/outline`
> 
> Move everything in `build` up a level and revert [06a70ad](https://github.com/NixOS/nixpkgs/commit/06a70adfa9c6f303306a595bb675465e40080311).

Actually this was my initial fix. I would prefer reverting changes in 06a70ad if we can resolve the email notification issue in some other way.

**Comment by weiren2** [comment]
I spent some more time on this issue. Commit 06a70ad moves the public assets (`public`) to `build` directory (`$out/share/outline/build`) during the install phase. This should fix the issue with [path to the email image](https://github.com/outline/outline/blob/21ce145f67b26cd7c4493258c65b5d1731b23a79/server/emails/mailer.tsx#L87), which assumes that `public` directory is accessible in the current working directory (CWD is `$out/share/outline/build` on NixOS). However, moving the location of  `public` breaks the existing [routing/path of static images](https://github.com/outline/outline/blob/21ce145f67b26cd7c4493258c65b5d1731b23a79/server/routes/index.ts#L32), which uses `__dirname` and some relative path (i.e., it assumes `public` is accessible under `$out/share/outline`).

### TLDR:
In the end, I just reverted the changes of 06a70ad and create a symlink of `public` under the `build` directory. The links to static images are working again.

### Questions:
 - Hopefully this can also address the original issue in 06a70ad. @xanderio Could you help test if this PR actually fixes the problem?
 - Since I'm still a newbie in the Nix world, I'm not sure if using symlinks when packaging is a good practice or not. Please let me know if there are better ways.

**Comment by weiren2** [comment]
> Can we change:
> 
> https://github.com/NixOS/nixpkgs/blob/042598f1989709751cab749a94dffdba9e926593/nixos/modules/services/web-apps/outline.nix#L784
> 
> To `${cfg.package}/share/outline`
> 
> Move everything in `build` up a level and revert [06a70ad](https://github.com/NixOS/nixpkgs/commit/06a70adfa9c6f303306a595bb675465e40080311).

Also this suggested fix didn't work unfortunately (without patching the upstream code). The `public` directory is assumed to be [three level up](https://github.com/outline/outline/blob/21ce145f67b26cd7c4493258c65b5d1731b23a79/server/routes/index.ts#L32) from `$build/server/routes`. So to move contents of `build` up one level, we still need to somehow link/duplicate `public` one level up or patch the upstream code.

That's why I decided to use symlink and try to keep the changes minimal.

### Alternative 
Following this idea (changing the working dir), another possible fix is to stick to the setup of [upstream docker](https://github.com/outline/outline/blob/21ce145f67b26cd7c4493258c65b5d1731b23a79/Dockerfile) as close as possible. In short:

 - We copy the relevant directory and files. In `nixpkgs/pkgs/servers/web-apps/outline/default.nix`:
```
# installPhase:
...
    mkdir -p $out/bin $out/share/outline
    mv build public node_modules server package.json $out/share/outline/
...
```
- and then change the working directory in `nixpkgs/nixos/modules/services/web-apps/outline.nix`
```
# serviceConfig:
...
 WorkingDirectory = "${cfg.package}/share/outline"; 
...
```
Let me know if we want to implement in this way.

**Comment by NickCao** [comment]
I don't have strong opinions here (not actually an outline user), cc @xanderio for ideas.

**Comment by xanderio** [comment]
Let's go with the symlink option, that was in fact the first way I patched the package for me internally. I wasn't aware that the `public` directory needs this specially placement in the filesystem tree and as such I switch from a symlink to moving the directory.

I just didn't notices that static images were broken after moving the `public` directory. (Presumably my browsers cache got in the way)

**Comment by weiren2** [comment]
Great! Hopefully this PR can be merged soon or get incorporated into the next version bump.

**Comment by NickCao** [comment]
diff LGTM, did't test though, and please squash the commits.

**Comment by weiren2** [comment]
Commits squashed into one (5f29a56).

**Review by NickCao — APPROVED** [review]
